service/keychain/local_keychain/local_keychain.go

```diff
@@ -0,0 +1,130 @@
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package local_keychain
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/awslabs/soci-snapshotter/proto"
+	"github.com/awslabs/soci-snapshotter/service/resolver"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/reference"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+var (
+	localKeychainPort = flag.Int("local_keychain_port", 0,
+		"Port on which to expose the local_keychain gRPC service that accepts username/password credentials for private images. If 0, the local_keychain service is not started/exposed.")
+)
+
+type credentials struct {
+	username   string
+	password   string
+	expiration *time.Time
+}
+
+type keychain struct {
+	cache map[string]credentials
+	proto.UnimplementedLocalKeychainServer
+}
+
+var singleton *keychain
+var lock = &sync.Mutex{}
+
+func (kc *keychain) init() {
+	if *localKeychainPort == 0 {
+		log.G(context.Background()).Info("no local_keychain_port specified, not starting local_keychain gRPC server")
+		return
+	}
+
+	lis, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", *localKeychainPort))
+	if err != nil {
+		log.G(context.Background()).Fatalf("failed to listen: %v", err)
+	} else {
+		log.G(context.Background()).Infof("started local keychain server on localhost:%d", *localKeychainPort)
+	}
+	var opts []grpc.ServerOption
+	grpcServer := grpc.NewServer(opts...)
+	proto.RegisterLocalKeychainServer(grpcServer, kc)
+	reflection.Register(grpcServer)
+	go grpcServer.Serve(lis)
+}
+
+func (kc *keychain) PutCredentials(ctx context.Context, req *proto.PutCredentialsRequest) (res *proto.PutCredentialsResponse, err error) {
+	if req.ImageName != "" && req.Credentials != nil {
+		var expirationTime *time.Time
+		if req.ExpiresInSeconds > 0 {
+			timeout := time.Now().Add(time.Duration(req.ExpiresInSeconds) * time.Second)
+			expirationTime = &timeout
+		}
+		kc.cache[req.ImageName] = credentials{
```

Comment:
you need locking around this map since it's being accessed & modified concurrently

Reply:
Oof, of course. Fixed.